### PR TITLE
Scheduler currentTime indication implementation

### PR DESCRIPTION
**Pull request by Yarovaya**

### CODE REVIEW

File: js/ui/scheduler/ui.scheduler.currentTimeShader.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    DATE_TIME_SHADER_BOTTOM_CLASS = "dx-scheduler-date-time-shader-bottom";
+
+
+var currentTimeShader = Class.inherit({
+    render: function(workspace, isVertical) {
+        this._workspace = workspace;
+        this._$container = workspace._dateTableScrollable.content();
+
+        if(isVertical) {
+            this._renderVerticalShader();
```
Comment (lukyanovas) [inline js/ui/scheduler/ui.scheduler.currentTimeShader.js]: I think what you should separate shader to vertical and horizontal

File: js/ui/scheduler/ui.scheduler.timeline.js [hunk 2]
```diff
@@ -159,12 +159,54 @@ var SchedulerTimeline = SchedulerWorkSpace.inherit({
...
+        this._renderDateTimeIndication();
+        this._setIndicationUpdateInterval();
+
         this._$sidebarTable.appendTo(this._sidebarScrollable.content());
 
         this._setGroupHeaderCellsHeight();
         this._applyCellTemplates(groupCellTemplates);
     },
 
+    _getIndicationWidth: function() {
```
Comment (lukyanovas) [inline js/ui/scheduler/ui.scheduler.timeline.js]: remove _

File: js/ui/scheduler/ui.scheduler.agenda.js [hunk 3]
```diff
@@ -308,6 +308,10 @@ var SchedulerAgenda = SchedulerWorkSpace.inherit({
     _attachEvents: noop,
     _cleanCellDataCache: noop,
 
+    _needRenderDateTimeIndication: function() {
```
Comment (lukyanovas) [inline js/ui/scheduler/ui.scheduler.agenda.js]: remove _

File: js/ui/scheduler/ui.scheduler.timeline.js [hunk 4]
```diff
@@ -4,9 +4,10 @@ var $ = require("../../core/renderer"),
     noop = require("../../core/utils/common").noop,
     extend = require("../../core/utils/extend").extend,
     registerComponent = require("../../core/component_registrator"),
-    SchedulerWorkSpace = require("./ui.scheduler.work_space"),
+    SchedulerWorkSpace = require("./ui.scheduler.work_space.indicator"),
```
Comment (SergeyAlexeev) [inline js/ui/scheduler/ui.scheduler.timeline.js:7]: It looks like you are requiring an indicator, not workspace. `ui.scheduler.work_space.indicator` seems ambiguous.

File: js/ui/scheduler/ui.scheduler.work_space.indicator.js [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+    _needRenderDateTimeIndicator: function() {
+        var now = this._getToday(),
+            endViewDate = dateUtils.trimTime(this.getEndViewDate());
+
+        return dateUtils.dateInRange(now, this._firstViewDate, new Date(endViewDate.getTime() + toMs("day")));
+    },
+
+    needRenderDateTimeIndication: function() {
+        var now = this._getToday();
```
Comment (SergeyAlexeev) [inline js/ui/scheduler/ui.scheduler.work_space.indicator.js]: var today = ...?

File: js/ui/scheduler/ui.scheduler.work_space.indicator.js [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+                }
+            }
+        }
+    },
+
+    _setIndicationUpdateInterval: function() {
+        if(!this.option("showCurrentTimeIndicator") || this.option("indicatorUpdateInterval") === 0) {
+            return;
+        }
+        var that = this;
```
Comment (SergeyAlexeev) [inline js/ui/scheduler/ui.scheduler.work_space.indicator.js]: `that` looks like 'old-school approach':) Use `bind()` instead.

File: js/ui/scheduler/ui.scheduler.work_space.indicator.js [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+            width = difference * this.getCellWidth();
+
+        return maxWidth < width ? maxWidth : width;
+    },
+
+    _getRoundedCellWidth: function() {
+        var $row = this.$element().find("." + this._getDateTableRowClass()).eq(0),
+            width = 0,
+            $cells = $row.find("." + this._getDateTableCellClass()),
+            cellsCount = $cells.length;
```
Comment (SergeyAlexeev) [inline js/ui/scheduler/ui.scheduler.work_space.indicator.js]: cellsCount -> cellCount

File: js/ui/scheduler/ui.scheduler.work_space.indicator.js [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    _refreshDateTimeIndication: function() {
+        this._cleanDateTimeIndicator();
+        this._shader && this._shader.clean();
+        this._renderDateTimeIndication();
+    },
+
+    _isCurrentTime: function(date) {
+        if(this.option("showCurrentTimeIndicator") && this._needRenderDateTimeIndicator()) {
+            var now = this._getToday(),
+                result = false;
```
Comment (SergeyAlexeev) [inline js/ui/scheduler/ui.scheduler.work_space.indicator.js]: now -> today

File: js/ui/scheduler/ui.scheduler.work_space.indicator.js [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+    _refreshDateTimeIndication: function() {
+        this._cleanDateTimeIndicator();
+        this._shader && this._shader.clean();
+        this._renderDateTimeIndication();
+    },
+
+    _isCurrentTime: function(date) {
+        if(this.option("showCurrentTimeIndicator") && this._needRenderDateTimeIndicator()) {
+            var now = this._getToday(),
```
Comment (SergeyAlexeev) [inline js/ui/scheduler/ui.scheduler.work_space.indicator.js]: now -> today

File: js/ui/scheduler/ui.scheduler.work_space.indicator.js [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+        }
+    },
+
+    _getTimeCellClass: function(i) {
+        var startViewDate = this._getTimeCellDate(i),
+            cellClass = this.callBase(i);
+
+        if(this._isCurrentTime(startViewDate)) {
+            return cellClass + " " + TIME_PANEL_CURRENT_TIME_CELL_CLASS;
+        } else {
```
Comment (SergeyAlexeev) [inline js/ui/scheduler/ui.scheduler.work_space.indicator.js]: The `else` statement isn't required. Just return cellClass.

File: styles/widgets/base/scheduler.less [hunk 11]
```diff
@@ -442,6 +445,111 @@
     display: none;
 }
 
+.dx-scheduler-header-panel-cell.dx-scheduler-header-panel-current-time-cell {
+    box-shadow: inset 0px -2px 0px 0px @SCHEDULER_APPOINTMENT_BASE_COLOR;
```
Comment (SergeyAlexeev) [inline styles/widgets/base/scheduler.less]: Can we use cross-browser mixin here?

File: styles/widgets/base/scheduler.less [hunk 12]
```diff
@@ -442,6 +445,111 @@
...
+             background-color:  @SCHEDULER_TIME_INDICATOR_BACKGROUND_COLOR;
+             pointer-events: none;
+         }
+    }
+
+    &.dx-scheduler-work-space-both-scrollbar {
+        .dx-scheduler-date-time-shader,
+        .dx-scheduler-date-time-indicator,
+        .dx-scheduler-date-time-shader-all-day {
+            margin-left: 0;
```
Comment (SergeyAlexeev) [inline styles/widgets/base/scheduler.less:513]: Does it work correctly in rtl mode?

File: testing/tests/DevExpress.ui.widgets.scheduler/currentTimeIndicator.tests.js [hunk 13]
```diff
@@ -0,0 +1,613 @@
...
+
+    QUnit.test("Shader on allDayPanel should have correct height & width, Week view", function(assert) {
+        this.instance.option({
+            indicatorTime: new Date(2017, 8, 7, 12, 45),
+            showAllDayPanel: true,
+            allDayExpanded: false
+        });
+        var $element = this.instance.$element();
+
+        assert.roughEqual($element.find("." + SCHEDULER_DATE_TIME_SHADER_ALL_DAY_CLASS).eq(0).outerHeight(), 24, 1, "Indicator has correct height");
```
Comment (SergeyAlexeev) [inline testing/tests/DevExpress.ui.widgets.scheduler/currentTimeIndicator.tests.js:417]: Extract `$element.find("." + SCHEDULER_DATE_TIME_SHADER_ALL_DAY_CLASS).eq(0)` to a separate variable.

File: styles/widgets/base/scheduler.less (hunk 12 above)
Comment (Yarovaya) [reply styles/widgets/base/scheduler.less:513]: Indicator behaviour in rtl mode was fixed for day/week views. There are some problems in timelines https://isc.devexpress.com/Thread/WorkplaceDetails/T561264

**Review by lukyanovas — APPROVED** [review]
LGTM

**Pull request merged**